Use window manager frame on Linux

bjorn opened the pull request:
Also, hide the custom "Always stay on top" option, which did not work on GNOME and possibly other window managers. The window manager's native option to keep a window on top can now be used.

Closes #262

This is how it looks on Ubuntu:

**Before**
![notes-no-frame](https://user-images.githubusercontent.com/531764/77419006-03a12280-6dc8-11ea-9f02-4ec3c19c40bc.png)
**After**
![notes-native-frame](https://user-images.githubusercontent.com/531764/77419011-0439b900-6dc8-11ea-8a1f-ad5e15d56220.png)

The screenshots also shows some DPI issues, probably due to hardcoded sizes that are not adapting to the font size. I guess it would be nice to fix that for 1.5 as well.

I also noticed the "Hide Notes" action in the tray icon is not working for me. But that is also unrelated to this patch.

theshadowx commented:
It's better to leave this after 1.5.0

bjorn commented:
> It's better to leave this after 1.5.0

Why? Do you think it's better to leave the broken "Always stay on top" action in for 1.5, or do you have another suggestion about what to do for 1.5?

theshadowx commented:
@bjorn we just remove it from menu item for now
And in post 1.5.0 we will tacle the gui issues
What do you think ?

bjorn commented:
> @bjorn we just remove it from menu item for now
> And in post 1.5.0 we will tacle the gui issues
> What do you think ?

Well, you suggested we use the window manager frame instead to still have the keep-on-top state available and I'd prefer this to just hiding the action.

I don't care that much either way though. Maybe @nuttyartist can help make the call for 1.5.

nuttyartist commented:
Hey all :)

@bjor, What DE did you use while taking these screenshots?

I'm siding with @theshadowx. The native option ruins the aesthetics of Notes. I know though, that if we are to enable native decoration after 1.5, it will not look pretty either, but I would prefer to do some user research to see what Linux users think and to test on different DE's to check how it looks. Right now, my intuition is that people download our app, mainly on Linux, because they see someone finally cared for design and not merely about functionality. So let's wait with this till after 1.5.0.

BTW, excuse my delay in responses lately. I just finished a marathon on a new thing I'm building, so now I will have more time to dedicate for Notes.

theshadowx commented:
The dowloads from github shows that Notes is a also used in windows.
This is for only the version 1.0.0
![20200326_103623](https://user-images.githubusercontent.com/976569/77637689-f31ab480-6f4d-11ea-9ad0-5d1786b1026a.jpg)

nuttyartist commented:
> The dowloads from github shows that Notes is a also used in windows.
This is for only the version 1.0.0

Yes, Indeed. Most of our users are Windows users (because of the OS share in Google searches).

bjorn commented:
> The native option ruins the aesthetics of Notes. I know though, that if we are to enable native decoration after 1.5 it will not look pretty either, but I would prefer to do some user research to see what Linux users think and to test on different DE's to check how it looks. Right now, my intuition is that people download our app mainly on Linux because they see someone finally cared for design and not merely about functionality.

It's a matter of opinion. As a Linux user, I like the consistency with other applications regarding the window frame, the drop shadow (the custom one looks wrong) and the position of the buttons.

What about I try to find some time to turn this into a user option? That should also make it easier to do user research. :-)

Waqar144 commented:
> It's a matter of opinion. As a Linux user, I like the consistency with other applications regarding the window frame, the drop shadow (the custom one looks wrong) and the position of the buttons.

Me as well. Linux already provides a lot of options for making your windows look like whatever you want.

theshadowx commented:
> It's a matter of opinion. As a Linux user, I like the consistency with other applications regarding the window frame

absolutely !

> the drop shadow (the custom one looks wrong) and the position of the buttons.

Could you elaborate ?

Waqar144 commented:
> Could you elaborate ?

I assume he means "it looks out of place" and not like other windows.

nuttyartist commented:
> What about I try to find some time to turn this into a user option? That should also make it easier to do user research. :-)

I'm totally in for this idea. Sounds good to me. Make the current custom decoration default but have an option for native window frame for Linux users using the -> View->Native window frame. I made an issue #264.

nuttyartist commented:
@bjorn Can you implement #264, our last issue for v1.5?

bjorn commented:
@nuttyartist I can try to find some time for it, but not this week.

nuttyartist commented:
@bjorn Got some progress?

bjorn commented:
@nuttyartist I really wanted to get Tiled 1.4 out first and yesterday I finally finished that release. I'll see if I can get around to this soon now.

nuttyartist commented:
You have done it well! Just two things:

1. Seems like there are conflicting files. Can you resolve it?
2. When a user chooses to use the native window frame, Can you close the gap that was originally for the 3 traffic lights? Talking about the red part in the following image:

![Screen Shot 2020-06-22 at 1 49 14 PM](https://user-images.githubusercontent.com/16375940/85280070-193deb00-b490-11ea-8577-ffad896b908d.png)

nuttyartist commented:
I also appreciate that you make sure to save the option the user chose for next time a user starts the program. 👍

nuttyartist commented:
@bjorn Any progress?

bjorn commented:
@nuttyartist I've resolved the conflicts, fixed a small issue I noticed on Linux and removed the empty space above the search box when the native frame is enabled.

I've only removed the space above the search box, because removing it also above the note text field would result in inconsistent margins around the controls at the top (where they would be rather closer to the top than to the sides).

Note that before merging, this feature also still needs to be tested on Windows.

nuttyartist commented:
Awesome, @bjorn! I will make sure to test it tomorrow and then merge (: 👍

nuttyartist commented:
Okay, so you successfully resolved previous issues, but now the textEditor needs go up a bit up. The newNoteButton, editorDateLabel, trashButton, and dotsButton needs to align with lineEdit.
![Screen Shot 2020-07-01 at 8 55 49 AM](https://user-images.githubusercontent.com/16375940/86208377-1c646580-bb79-11ea-8054-a05e6dcfd9ab.png)

nuttyartist commented:
@bjorn Let me know when you finish this.

bjorn commented:
@nuttyartist As I wrote before:

> I've only removed the space above the search box, because removing it also above the note text field would result in inconsistent margins around the controls at the top (where they would be rather closer to the top than to the sides).

Let me know if you think the space should still be removed.

nuttyartist commented:
I'm sorry, I haven't noticed! Do you have a picture of that? Or a source I can test?

bjorn commented:
Comparison:

<img src="https://user-images.githubusercontent.com/531764/87253614-eed2c280-c47c-11ea-97be-d7d81d475cd0.png" width="400" alt="notes1" /> <img src="https://user-images.githubusercontent.com/531764/87253615-ef6b5900-c47c-11ea-85bd-bfc8c073cc59.png" width="400" alt="notes2" />

I'm a little undecided, but the one on the left seems cleaner to me. On the right, the margins to the left/right of the header do not match with the margin at the top.

nuttyartist commented:
It's definitely should be the right picture, where the elements are alinged with lineEdit. 

> On the right, the margins to the left/right of the header do not match with the margin at the top.

This is because of the native window. A solution might be, when native window is on, to create a spacer that will move down the entire elements a bit. So everything would still align properly.

bjorn commented:
> This is because of the native window. A solution might be, when native window is on, to create a spacer that will move down the entire elements a bit. So everything would still align properly.

So, to me, a possible solution is the version on the left. Another option might be to reduce the spacing on the left/right of the editor, such that it matches with the reduced space at the top.

When we use a spacer to move everything down, wouldn't we get the same situation we had initially at https://github.com/nuttyartist/notes/pull/263#issuecomment-647443116?

nuttyartist commented:
> So, to me, a possible solution is the version on the left. Another option might be to reduce the spacing on the left/right of the editor, such that it matches with the reduced space at the top.

I can't settle on unaligned elements. This is bad design.

> When we use a spacer to move everything down, wouldn't we get the same situation we had initially at #263 (comment)?

True. Do you think adding just a little bit would look better?

bjorn commented:
> I can't settle on unaligned elements. This is bad design.

What about the suggestion to reduce the spacers left/right of the text editor to make things align again? Like this:

<img src="https://user-images.githubusercontent.com/531764/87436075-d8e40f80-c5ec-11ea-9f39-7f55bbd2fdf7.png" width="500" alt="notes1" />

nuttyartist commented:
> What about the suggestion to reduce the spacers left/right of the text editor to make things align again? Like this:

Can you create a branch for this so I can test on my machine? And another one without reducing the margins?

bjorn commented:
> Can you please create a branch for this so I can test on my machine? And another one without reducing the margins?

I've pushed the version on the right in https://github.com/nuttyartist/notes/pull/263#issuecomment-657258113 to this PR, and I've created a branch in my repo called [`consistent-spacing-with-native-frame`](https://github.com/bjorn/notes/tree/consistent-spacing-with-native-frame) that also reduces the space on left and right of the text edit.

bjorn commented:
I just tested this on Windows and turning off the native frame is not working correctly:

<img src="https://user-images.githubusercontent.com/531764/87585965-19708580-c6e0-11ea-90e9-b965c860a4ef.png" width="500" alt="notes1" /> 

It only works after restart.

Also, I noticed the window buttons are always on the left. That is really weird on Windows.

nuttyartist commented:
> I've pushed the version on the right in #263 (comment) to this PR, and I've created a branch in my repo called consistent-spacing-with-native-frame that also reduces the space on left and right of the text edit.

Okay let's go with 'consistent-spacing-with-native-frame'. One problem, on 'consistent-spacing-with-native-frame', the textEditor didn't revert to the right margins after turning off the native window frame.

> I just tested this on Windows and turning off the native frame is not working correctly:
>
> <img src="https://user-images.githubusercontent.com/531764/87585965-19708580-c6e0-11ea-90e9-b965c860a4ef.png" width="500" alt="notes1" /> 
>
> It only works after restart.

Do you mean getting out of the native window doesn't happen instantly on Windows and only after restart? That's really weird. If this problem is consistent on other machines we should show a prompt asking the user if he wants to do it.

> Also, I noticed the window buttons are always on the left. That is really weird on Windows.

When native-window is on, can you hide these buttons like you do on linux/mac?

bjorn commented:
> Okay let's go with 'consistent-spacing-with-native-frame'. One problem, on 'consistent-spacing-with-native-frame', the textEditor didn't revert to the right margins after turning off the native window frame.

Yeah, I had explained that part in the commit message of dfc8d62d728b86b6f95482de34f228caba81823f. :-)

> Do you mean getting out of the native window doesn't happen instantly on Windows and only after restart? That's really weird. If this problem is consistent on other machines we should show a prompt asking the user if he wants to do it.

What happens it what I showed on the screenshot, where our own buttons become visible but the native frame does not disappear. This bug only happens on Windows. I'll see if I can fix it.

> When native-window is on, can you hide these buttons like you do on linux/mac?

Yes, that already happens.

nuttyartist commented:
> Yeah, I had explained that part in the commit message of dfc8d62. :-)

I don't understand, why isn't as simple as changing the stylesheet when unchecking the native window frame option?

> What happens is what I showed on the screenshot, where our own buttons become visible but the native frame does not disappear. This bug only happens on Windows. I'll see if I can fix it.

👍 

> Yes, that already happens. Edit: To be clear, I didn't mean the buttons were "always on", I meant to say that our buttons were "on the left", whereas on Windows these buttons are on the right side for all other applications.

Ah, ok. Yes, we should move it to the right. I'm going to open an issue for this.

bjorn commented:
> I don't understand, why isn't as simple as changing the stylesheet when unchecking the native window frame option?

I never said it wasn't simple, I just pointed out it still needs to be done. It needs some code reorganization because that stylesheet is set in a function that can only be called once. I didn't have time for this yet.

nuttyartist commented:
Ok 👍

nuttyartist commented:
Hey @bjorn, any progress?

nuttyartist commented:
@bjorn I just tested on Windows and can see what you mean about the native window frame not disappearing before doing a system restart. If you think it's an upstream bug or something that's hard to figure out, maybe we can just show a dialog after choosing to show native window frame. Something like: "After you do this action, In order to revert back to custom decoration, you will need to restart your system. Are you sure you want to do this? Yes/No.". What do you think?

bjorn commented:
> If you think it's an upstream bug or something that's hard to figure out, maybe we can just show a dialog after choosing to show native window frame. Something like: "After you do this action, In order to revert back to custom decoration, you will need to restart your system. Are you sure you want to do this? Yes/No.". What do you think?

Yeah, I still need to look into this but if we don't find a solution then such a dialog would work.

I've been rather busy so I had no time yet to continue this patch.

nuttyartist commented:
That's alright. I would love to release v1.5 till the end of the month, so if we can wrap this up by then it would be great.

nuttyartist commented:
@bjorn Can you add this code for the dialog so we will merge this?

```c++
QMessageBox msgBox;
msgBox.setText(tr("Warning: After performing this action, to revert to custom decoration you will need to restart your system."));
msgBox.setInformativeText(tr("Would you like to continue?"));
msgBox.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
msgBox.setDefaultButton(QMessageBox::No);
if  (msgBox.exec() != QMessageBox::Yes) {
    return;
}
```

nuttyartist commented:
Okay, merging this now. I'll add the dialog code afterward.

bjorn commented:
Alright, that's fine. Will see if I can still get around to the margin changes...

bjorn commented:
@nuttyartist That part:

> > I don't understand, why isn't as simple as changing the stylesheet when unchecking the native window frame option?
> 
> I never said it wasn't simple, I just pointed out it still needs to be done. It needs some code reorganization because that stylesheet is set in a function that can only be called once. I didn't have time for this yet.

nuttyartist commented:
Oh, okay!
